src/entities/permit2/allowanceTransfer.ts

```diff
@@ -0,0 +1,193 @@
+import { Address, TypedData, TypedDataDomain, hashTypedData } from 'viem';
+
+import {
+    MaxSigDeadline,
+    MaxOrderedNonce,
+    MaxAllowanceTransferAmount,
+    MaxAllowanceExpiration,
+} from './constants';
+import { permit2Domain } from './domain';
+
+export interface PermitDetails {
+    token: Address;
+    amount: bigint;
+    expiration: number;
+    nonce: number;
+}
+
+export interface PermitSingle {
+    details: PermitDetails;
+    spender: Address;
+    sigDeadline: bigint;
+}
+
+export interface Permit2Batch {
+    details: PermitDetails[];
+    spender: Address;
+    sigDeadline: bigint;
+}
+
+export type PermitSingleData = {
+    domain: TypedDataDomain;
+    types: TypedData;
+    values: PermitSingle;
+};
+
+export type PermitBatchData = {
+    domain: TypedDataDomain;
+    types: TypedData;
+    values: Permit2Batch;
+};
+
+const PERMIT_DETAILS = [
+    { name: 'token', type: 'address' },
+    { name: 'amount', type: 'uint160' },
+    { name: 'expiration', type: 'uint48' },
+    { name: 'nonce', type: 'uint48' },
+] as const;
+
+const PERMIT_TYPES = {
+    PermitDetails: PERMIT_DETAILS,
+    PermitSingle: [
+        { name: 'details', type: 'PermitDetails' },
+        { name: 'spender', type: 'address' },
+        { name: 'sigDeadline', type: 'uint256' },
+    ],
+} as const;
+
+const PERMIT_BATCH_TYPES = {
+    PermitDetails: PERMIT_DETAILS,
+    PermitBatch: [
+        { name: 'details', type: 'PermitDetails[]' },
+        { name: 'spender', type: 'address' },
+        { name: 'sigDeadline', type: 'uint256' },
+    ],
+} as const;
+
+function isPermit(permit: PermitSingle | Permit2Batch): permit is PermitSingle {
```

Comment:
Never seen this type of thing before: `permit is PermitSingle` seems like its just another way of saying `boolean` or is there a specific reason for it?